src/Microsoft.OData.Edm/VocabularyAnnotationCache.cs

```diff
@@ -0,0 +1,41 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.OData.Edm.Vocabularies;
+
+namespace Microsoft.OData.Edm
+{
+    /// <summary>
+    /// Cache for storing a model's vocabulary annotations for quicker retrieval
+    /// </summary>
+    internal class VocabularyAnnotationCache
+    {
+        private readonly ConcurrentDictionary<IEdmVocabularyAnnotatable, IEnumerable<IEdmVocabularyAnnotation>> annotationsCache =
+            new ConcurrentDictionary<IEdmVocabularyAnnotatable, IEnumerable<IEdmVocabularyAnnotation>>();
+
+        /// <summary>
+        /// Adds vocabulary annotations for the specified <paramref name="element"/> to the cache.
+        /// </summary>
+        /// <param name="element">The element for which to cache annotations.</param>
+        /// <param name="annotations">The vocabulary annotations to cache</param>
+        public void AddVocabularyAnnotations(IEdmVocabularyAnnotatable element, IEnumerable<IEdmVocabularyAnnotation> annotations)
+        {
+            annotationsCache[element] = annotations;

```

Comment:
annotationsCache.TryAdd() is more efficient than using the assignment operator, because if the value already exists it will proceed, and not acquire internal locks just to assign the same annotations again.

Reply:
Thanks. That's a good point.